src/main/python/id_contexts/up_student_card.py

```diff
@@ -0,0 +1,109 @@
+
+import re
+from id_contexts.id_context import IDContext, FieldType, LineType
+from hutts_utils.hutts_logger import logger
+
+
+class UPStudentCard(IDContext):
+    """
+    A class that represents an ID context for a University of Pretoria ID card.
+    """
+    def __init__(self):
+        """
+        Initialises the UPStudentCard object.
+        """
+        # Logging for debugging purposes.
+        logger.debug('Initialising %s...' % type(self).__name__)
+        # Specify initial list of contexts for string image_processing when populating
+        # the ID information dictionary to send as output.
+        match_contexts = [{
+            'field': 'identity_number',
+            'find': None,
+            'field_type': FieldType.NUMERIC_ONLY,
+            'line_type': LineType.UNTITLED_ADJACENT,
+            'multi_line': False
+        }, {
+            'field': 'surname',
+            'find': None,
+            'field_type': FieldType.TEXT_ONLY,
+            'to_uppercase': False,
+            'line_type': LineType.UNTITLED_ADJACENT,
+            'multi_line': False
+        }, {
+            'field': 'names',
+            'find': None,
+            'field_type': FieldType.TEXT_ONLY,
+            'to_uppercase': True,
+            'line_type': LineType.TITLED_NEWLINE,
+            'multi_line': False,
+        }]
+        # Initialise parent
+        IDContext.__init__(self, match_contexts)
+
+    def _dictify(self, match_contexts, id_string, barcode_data, fuzzy_min_ratio, max_multi_line):
+        """
+        This function is responsible for generating a dictionary object containing the relevant ID information,
+        such as names, surname, ID number, etc., from a given input string containing said relevant information.
+        In this particular ID context, the information is sparse and is mainly intended for demonstration purposes.
+
+        Authors:
+            Jan-Justin van Tonder
+
+        Args:
+            match_contexts (list): A list of dictionaries that contain the contextual information used in the process
+                of retrieving field values from the OCR output string - not particularly useful for this ID context.
+            id_string (str): A string containing some ID information.
+            barcode_data (dict, Optional): A dictionary object containing information extracted from a barcode.
+            fuzzy_min_ratio (float): The threshold ratio for a minimum, acceptable ratio of fuzziness when comparing
+                two strings.
+            max_multi_line (int): Specifies the maximum number of lines that is to be extracted from fields that are
+                noted as running onto multiple lines.
+
+        Returns:
+            (dict): A dictionary object containing the relevant, extracted ID information.
+        """
+        id_info = {}
+        regexp = re.compile(r'[0-9]{6,10}')
+        # Check if barcode data is available.
+        if barcode_data and barcode_data['identity_number']:
```

Comment:
The problem with setting the default value of the barcode_data argument to an empty dictionary is that the dictionary is mutable. This can cause problems due to the way python handles default arguments and is considered bad practice since in the event that the called function makes changes to the default argument value, this change is persisted across all future function calls.

I could, however, do an assignment within the function but then more checks would have to be performed.

Reply:
Ok I understand.